### PR TITLE
feat(x/ecocredit): v1 basket query proto

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

Adds query proto defs for baskets

Ref: #732

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    // Baskets lists all baskets in the ecocredit module.
+    rpc Baskets(QueryBasketsRequest) returns (QueryBasketsResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/v1alpha1/baskets";
+    }
+
+    // BasketBalanceById lists the balance of each credit batch held by the basket.
+    rpc BasketBalanceById(QueryBasketBalanceByIdRequest) returns (QueryBasketBalanceByIdResponse) {
```
Comment (aaronc) [inline proto/regen/ecocredit/basket/v1/query.proto:27]: Not sure we need this

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        option (google.api.http).get =
+            "/regen/ecocredit/v1alpha1/baskets/{basket_id}/balance";
+    }
+
+    // BasketBalanceByDenom lists the balance of each credit batch held by the basket.
+    rpc BasketBalanceByDenom(QueryBasketBalanceByDenomRequest) returns (QueryBasketBalanceByDenomResponse) {
```
Comment (aaronc) [inline proto/regen/ecocredit/basket/v1/query.proto:33]:
```suggestion
    rpc BasketBalancesByDenom(QueryBasketBalanceByDenomRequest) returns (QueryBasketBalanceByDenomResponse) {
```

**Review by aaronc — COMMENTED** [review]
We should have a basket balance by denom query that returns the balance for a single basket denom + batch denom pair

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 1 above)
Comment (technicallyty) [reply proto/regen/ecocredit/basket/v1/query.proto]: removed

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    string basket_denom = 1;
+
+    // batch_denom is the denom of the credit batch.
+    string batch_denom = 2;
+}
+
+// QueryBasketBalanceByDenomResponse is the Query/BasketBalanceByDenom response type.
+message QueryBasketBalanceByDenomResponse {
+    // balance is the balance of a specific credit batch in the basket.
+    BasketBalance balance = 1;
```
Comment (aaronc) [inline proto/regen/ecocredit/basket/v1/query.proto:96]: can't we just have a single integer string here?

**Review by aaronc — APPROVED** [review]

**Comment by aaronc** [comment]
why so much generated code changed?

**Comment by technicallyty** [comment]
> why so much generated code changed?

not sure - maybe i have different versions of protoc-gen-*'s?

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 4]
```diff
@@ -0,0 +1,97 @@
+syntax = "proto3";
+
+package regen.ecocredit.basket.v1;
+
+option go_package = "github.com/regen-network/regen-ledger/x/ecocredit/basket/v1";
```
Comment (ryanchristo) [inline proto/regen/ecocredit/basket/v1/query.proto:5]: Are we needed to remove `/v1` now? @robert-zaremba

**Review by ryanchristo — APPROVED** [review]
One question, otherwise looks good.

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 4 above)
Comment (robert-zaremba) [reply proto/regen/ecocredit/basket/v1/query.proto:5]: yes

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    // Basket queries one basket by denom.
+    rpc Basket(QueryBasketRequest) returns (QueryBasketResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/v1alpha1/baskets/{basket_denom}";
+    }
+
+    // Baskets lists all baskets in the ecocredit module.
+    rpc Baskets(QueryBasketsRequest) returns (QueryBasketsResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/v1alpha1/baskets";
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/basket/v1/query.proto:23]: I  think we should change this to  either: 
1. /regen/ecocredit/v1/baskets
2. /regen/ecocredit/baskets/v1

Comment (technicallyty) [reply proto/regen/ecocredit/basket/v1/query.proto:23]: i'll do 1. to keep in alignment with how we usually do it

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    // BasketBalances lists the balance of each credit batch in the basket.
+    rpc BasketBalances(QueryBasketBalancesRequest) returns (QueryBasketBalancesResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/v1alpha1/baskets/{basket_denom}/balances";
+    }
+
+    // BasketBalanceByDenom queries the balance of a specific credit batch in the basket.
+    rpc BasketBalanceByDenom(QueryBasketBalanceByDenomRequest) returns (QueryBasketBalanceByDenomResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/v1alpha1/baskets/{basket_denom}/balances/{batch_denom}";
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/basket/v1/query.proto]:
```suggestion
            "/regen/ecocredit/v1alpha1/baskets/{basket_denom}/balance/{batch_denom}";
```

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 5 above)
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/query.proto:23]: 2 is more consistent

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 4 above)
Comment (technicallyty) [reply proto/regen/ecocredit/basket/v1/query.proto]: fixed

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 5 above)
Comment (technicallyty) [reply proto/regen/ecocredit/basket/v1/query.proto]: okay edited to be more aligned with the package paths, how does it look now?

**Review by robert-zaremba — APPROVED** [review]

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    // BasketBalances lists the balance of each credit batch in the basket.
+    rpc BasketBalances(QueryBasketBalancesRequest) returns (QueryBasketBalancesResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/basket/v1/baskets/{basket_denom}/balances";
+    }
+
+    // BasketBalanceByDenom queries the balance of a specific credit batch in the basket.
+    rpc BasketBalanceByDenom(QueryBasketBalanceByDenomRequest) returns (QueryBasketBalanceByDenomResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/basket/v1/baskets/{basket_denom}/balances/{batch_denom}";
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/basket/v1/query.proto:35]:
```suggestion
            "/regen/ecocredit/basket/v1/baskets/{basket_denom}/balance/{batch_denom}";
```

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 5 above)
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/query.proto]: let's do 2 please, it lines up with the proto package path and i think will be the general way we do this

Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/query.proto]: LGTM now 👍

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 7 above)
Comment (technicallyty) [reply proto/regen/ecocredit/basket/v1/query.proto:35]: i would prefer to keep it this way to align with RESTful resource conventions. we already have the `balances` collection in the previous query so would prefer to keep it the way it is.

ref: https://restfulapi.net/resource-naming/ - 1.1. Singleton and Collection Resources

**Comment by aaronc** [comment]
let's fix proto lint

File: proto/regen/ecocredit/basket/v1/query.proto (hunk 7 above)
Comment (robert-zaremba) [reply proto/regen/ecocredit/basket/v1/query.proto:35]: OK, thanks for the link

File: proto/regen/ecocredit/basket/v1/query.proto [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    // BasketBalances lists the balance of each credit batch in the basket.
+    rpc BasketBalances(QueryBasketBalancesRequest) returns (QueryBasketBalancesResponse) {
+        option (google.api.http).get =
+            "/regen/ecocredit/basket/v1/baskets/{basket_denom}/balances";
+    }
+
+    // BasketBalanceByDenom queries the balance of a specific credit batch in the basket.
+    rpc BasketBalanceByDenom(QueryBasketBalanceByDenomRequest) returns (QueryBasketBalanceByDenomResponse) {
```
Comment (aaronc) [inline proto/regen/ecocredit/basket/v1/query.proto:33]: why not just
```suggestion
    rpc BasketBalance(QueryBasketBalanceByDenomRequest) returns (QueryBasketBalanceByDenomResponse) {
```

Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/query.proto:33]: auto-merge disabled pending whether we change this

Comment (technicallyty) [reply proto/regen/ecocredit/basket/v1/query.proto]: ok changed it

**Pull request merged**